### DISCUSSION TITLE
Adding a new incompressible fluid

### DISCUSSION THREAD

**Discussion by StartCodeLearner** [CoolProp/CoolProp#2353] [Q&A]
Hello, everyone, I have a question about adding an incompressible fluid. I have tried various methods, but still haven't worked not. 
1)Try reading from a json file, modify it then reload them, following the instructions: http://www.coolprop.org/coolprop/HighLevelAPI.html#adding-fluids
But errors for these two functions:
jj = CP.get_fluid_param_string('INCOMP::NaK', 'json')
AS = CP.AbstractState('HEOS', 'INCOMP::NaK')

2) I have made the JSON file for the new fluid following the instructions: http://www.coolprop.org/fluid_properties/Incompressibles.html#adding-new-fluids
How to add it to the code without recompiling the source code

3) Another question about the CP.set_config_bool(CP.DONT_CHECK_PROPERTY_LIMITS, True)
The config file has been changed but it doesn't work for incompressible fluids like INCOMP::NaK. 

Thanks in advance.

**Comment by ibell** [comment]
1. The adding of fluids in that way is only available for the other kinds of models, not the incompressible ones. In your example, it should read: ``CP.AbstractState('INCOMP','NaK')`` .  `` 'INCOMP'`` is the backend and ``'NaK'`` is the fluid name
2. You need to recompile (I think, right @jowr)?
3. Can you show a complete runnable example?

**Reply by StartCodeLearner** [reply]
hello, ibell, thanks for your help. The Tin for NaK (a kind of salt) is 573.15 K. I want to extend the calculation below to 290C.
Here is the code following your previous answers

 ```

import json
import CoolProp.CoolProp as CP

fluid='INCOMP::NaK'
#load the state
AS=CP.AbstractState('INCOMP','NaK')
CP.set_config_bool(CP.OVERWRITE_FLUIDS, True)

#Get the fluid information in JSON format
#for loaded fluid
jbase = json.loads(AS.fluid_param_string("JSON"))[0]
jbase["Tmin"] = 553.13 #set the Tin=553.15 K
CP.add_fluids_as_JSON('INCOMP', json.dumps(jbase))

#Here is the Test
density_at_290C = CP.PropsSI('D', 'P', 0.1E6, 'T', 290 + 273.15, 'INCOMP::NaK')
print(f'density_at_290C={density_at_290C}')

```

**Reply by ianhbell** [reply]
So that's your desired code, but it doesn't work as far as I can tell, right? Because partly at least JSON is not a valid option for fluid_param_string for the INCOMP backend

**Comment by StartCodeLearner** [comment]
Yes.That's right, Is there any simple way to add or modify new incomp fluid without compilation?



---- Replied Message ----
| From | Ian ***@***.***> |
| Date | 02/28/2024 07:04 |
| To | ***@***.***> |
| Cc | ***@***.***>***@***.***> |
| Subject | Re: [CoolProp/CoolProp] Adding a new incompressible fluid (Discussion #2353) |

So that's your desired code, but it doesn't work as far as I can tell, right? Because partly at least JSON is not a valid option for fluid_param_string for the INCOMP backend

**Comment by ibell** [comment]
Currently, no. But if you wanted to write a PR to add this capability it would be welcomed.

**Reply by caherrera97** [reply]
Dear Dr. Bell, I too am looking to add an incompressible fluid to CoolProp.  Has anything changed since your last reply here?

**Reply by ibell** [reply]
No - somebody needs to step up and get involved and add this capability if people are interested. CoolProp is a community project.
